README.md
# timecutrust

[![Build Status](https://travis-ci.org/bogdad/timecutrust.svg?branch=master)](https://travis-ci.org/bogdad/timecutrust)

quickly find a time position in a log file (like nginx access)
using binary search, and then output all the lines newer than it

# usage
```
./target/debug/timecutrust

Usage: timecutrust [options] 'beg-time' file

Options:
    -r, --regexp REGULAR_EXPRESSION
          like ^\[(\d{4})-(\d{2})-(\d{2})\s(\d{2}):(\d{2}):(\d{2})\]
          default - ^\[(\d{4})-(\d{2})-(\d{2})\s(\d{2}):(\d{2}):(\d{2})\]
    -n, --nginx         run on nginx access.log with pattern like
                        '[28/Dec/2015:06:26:08'
                    
    -h, --help          print this help menu

beg-time:
    Time of start in the format specified in -r

file:
   Logfile to parse

examples:
timecutrust '[2015-12-28 20:37:25]' ./sample_g2.log
```

debian/changelog
timecutrust (0.1.0) unstable; urgency=low

  * initial debianization

 -- Vladimir Shakhov <vshakhov@yandex-team.ru>  Tue, 28 Jun 2016 15:24:30 +0300


src/datetimes.rs
use regex::Regex;
use chrono::format::Fixed;
use chrono::format::Item;
use chrono::format::parsed::*;
use chrono::*;

pub fn init(r: &str) -> Regex {
    let re = Regex::new(r).unwrap();
    re
}

pub fn monthIndex(line: &str) -> u32 {
    let mut parsed: Parsed = Parsed::new();
    //println!("{}", line);
    let items = vec![Item::Fixed(Fixed::ShortMonthName)];
    let items_iter = items.into_iter();
    format::parse(& mut parsed, line, items_iter);
    parsed.month.unwrap()
}

pub fn parse(re: &Regex, line: &str) -> Option<DateTime<UTC>> {
    let caps_o = re.captures(line);
    let caps = match caps_o {
        None => return None,
        Some(c) => c
    };
    let year = match caps.name("year") {
        None => return None,
        Some(y) => y.parse::<i32>().expect("could not parse year")
    };
    let mon = match caps.name("monthname")
        .map(|mn| monthIndex(mn))
        .or(caps.name("month").map(|t| t.parse::<u32>().expect("could not parse month")))
         {
        None => return None,
        Some(m) => m
    };
    let day = match caps.name("day") {
        None => return None,
        Some(d) => d.parse::<u32>().expect("could not parse day")
    };
    let hh = match caps.name("hour") {
        None => return None,
        Some(hh) => hh.parse::<u32>().expect("could not parse hour")
    };
    let mm = match caps.name("minute") {
        None => return None,
        Some(mm) => mm.parse::<u32>().expect("could not parse minute")
    };
    let ss = match caps.name("second") {
        None => return None,
        Some(ss) => ss.parse::<u32>().expect("could not parse second")
    };
    let dt = UTC.ymd(year, mon, day).and_hms(hh, mm, ss);
    Some(dt)
}

#[test]
fn test_sample_regexp() {
    let re = init(r"^\[(?P<year>\d{4})-(?P<month>\d{2})-(?P<day>\d{2})\s(?P<hour>\d{2}):(?P<minute>\d{2}):(?P<second>\d{2})\]");
    let ti = parse(&re, "[2015-12-28 20:37:25]");
    assert_eq!(Some(2015), ti.map(|t| t.year()));
}

#[test]
fn test_nginx_regexp() {
    let re = init(r"^\[(?P<day>\d{2})/(?P<monthname>\p{L}*)/(?P<year>\d{4}):(?P<hour>\d{2}):(?P<minute>\d{2}):(?P<second>\d{2})");
    let ti = parse(&re, "[28/Dec/2015:20:37:25");
    assert_eq!(Some(2015), ti.map(|t| t.year()));
    assert_eq!(Some(12), ti.map(|t| t.month()));
}


src/binary_search.rs
//  1   2   3  4  5 6 7 8 9 10
//  -4 -3  -2 -1  0 1 2 3 4 5
pub fn binary_search<'a, P>(pi_beg: u64, pi_end: u64, predicate: &'a mut P) -> u64
    where P : FnMut(u64) -> i64 {
    let mut i_beg = pi_beg;
    let mut i_end = pi_end;
    let pval = predicate(i_beg);
    if pval == 0 {
        return i_beg;
    }
    while i_beg <= i_end {
        let mid = i_beg + (i_end-i_beg)/2;
        let pval = predicate(mid);
        //println!("{:?} {:?} {:?} {:?} \n", i_beg, i_end, mid, pval);
        if pval == 0 {
            return  mid;
        } else if pval < 0 {
            if mid < i_end {
                i_beg = mid + 1;
            } else {
                break;
            }
        } else {
            if mid > 0 {
                i_end = mid - 1;
            } else {
                break;
            }
        }
    };
    i_beg
}


struct EqPred {
    item : i64
}

impl EqPred {
    fn new(item:i64) -> EqPred {
        EqPred{item: item}
    }
    fn call_inner(&self, pos:(u64,)) -> i64 {
        let (i,) = pos;
        let z:i64 = self.item.checked_add(i as i64).unwrap();
        z
    }
}

impl FnOnce<(u64,)> for EqPred {
    type Output = i64;
    extern "rust-call" fn call_once(self, pos: (u64,)) -> i64 {
        self.call_inner(pos)
    }
}

impl FnMut<(u64,)> for EqPred {
    extern "rust-call" fn call_mut(& mut self, pos: (u64,)) -> i64 {
        self.call_inner(pos)
    }
}


#[test]
fn test_binary_search() {
    let mut eq2 = EqPred::new(-2);
    let mut eq9 = EqPred::new(-9);
    let mut eq10 = EqPred::new(-10);
    let mut eq_1 = EqPred::new(1);
    assert_eq!(binary_search(0, 9, &mut eq2), 2);
    assert_eq!(binary_search(0, 9, &mut eq9), 9);
    assert_eq!(binary_search(0, 9, &mut eq10), 9);
    assert_eq!(binary_search(0, 9, &mut eq_1), 0);
}



src/textfileutils.rs
use std::io::{SeekFrom, BufReader, Cursor};
use std::io::prelude::*;

pub fn get_first_line_after<'a, R: Read + Seek>(reader: &mut BufReader<R>, from: u64) -> Option<String> {
    find_new_line_pos(reader, from as usize)
}

const SIZE: usize = 512;
// it is able to find strings up-to 1024 elements wide
fn find_new_line_pos<'a, R: Read + Seek>(reader: &mut BufReader<R>, from: usize) -> Option<String> {
    let before = match from.checked_sub(SIZE) {
        None => 0,
        Some(x) => x
    };
    reader.seek(SeekFrom::Start(before as u64)).ok().expect(&format!("seek failed {}", before));
    let mut buf: [u8;2*SIZE] = [0; 2*SIZE];
    let len = reader.read(&mut buf).unwrap();
    if len == 0 {
        return None;
    }
    let bufs = &buf[0..len];
    //print!("{:?}", String::from_utf8_lossy(bufs));
    let last_before = bufs.iter().enumerate().rposition(|(i, x)| *x==b'\n' && (i + before) < from);
    let last_after = bufs.iter().enumerate().position(|(i, x)| *x==b'\n' && (i + before) >= from);
    //println!("from {} before {} lb {:?} la {:?}", from, before, last_before, last_after);
    let str_before = match last_before {
        None => String::from_utf8_lossy(&buf[0..from-before]).into_owned(),
        Some(pos) => String::from_utf8_lossy(&buf[pos+1..from-before]).into_owned()
    };
    let str_after = match last_after {
        None => String::from_utf8_lossy(&buf[from-before..len]).into_owned(),
        Some(pos) => String::from_utf8_lossy(&buf[from-before..pos]).into_owned()
    };
    Some(str_before + &str_after)
}

#[test]
fn find_new_line_pos_works() {
    let data = String::from("some\nother\nline");
    let bytes = data.as_bytes();
    let mut test_data = BufReader::new(Cursor::new(bytes));
    {
        let pos = find_new_line_pos(&mut test_data, 5);
        assert_eq!(Some("other".to_string()),pos);
    }
    {
        let pos = find_new_line_pos(&mut test_data, 1);
        assert_eq!(Some("some".to_string()), pos)
    }
    {
        let pos = find_new_line_pos(&mut test_data, 4);
        assert_eq!(Some("some".to_string()), pos);
    }
    {
        let pos = find_new_line_pos(&mut test_data, 10);
        assert_eq!(Some("other".to_string()), pos);
    }
    {
        let pos = find_new_line_pos(&mut test_data, 11);
        assert_eq!(Some("line".to_string()), pos);
    }
}


src/main.rs
#![feature(unboxed_closures, fn_traits, core)]
extern crate regex;
extern crate chrono;
extern crate getopts;

use std::marker::PhantomData;
use std::io::{self, BufReader, Lines, Write};
use std::fs::File;
use std::fs;
use std::io::prelude::{Seek, Read, BufRead};
use std::io::{SeekFrom, Error, ErrorKind};
use std::env;

use regex::Regex;
use chrono::UTC;
use chrono::{DateTime, Duration};

use getopts::Options;

mod binary_search;
mod regexps;
mod textfileutils;
mod datetimes;



struct FilePredicate<'a, R: 'a + Read + Seek> {
    file: BufReader<R>,
    len: u64,
    re: regex::Regex,
    b_time: DateTime<UTC>,
    phantom: PhantomData<&'a BufReader<R>>
}

impl<'a, R: 'a + Read + Seek> FilePredicate<'a, R> {
    fn new(file: BufReader<R>, len: u64, re: Regex, b_time: DateTime<UTC>) -> FilePredicate<'a, R> {
        FilePredicate{ file: file, len: len, re: re, b_time: b_time, phantom: PhantomData }
    }

    fn call_inner_mut(&mut self, args: (u64,)) -> i64 {
        let (pos,) = args;
        let mut cur_pos = pos;
        let mut line_timestamp: i64 = -1;
        loop {
            let line_o = textfileutils::get_first_line_after(& mut self.file, cur_pos);
            if line_o.is_none() {
                break
            }
            let line = line_o.unwrap();
            let line_time_o = datetimes::parse(&self.re, &line);
            if line_time_o.is_some() {
                line_timestamp = line_time_o.unwrap().timestamp();
                break;
            } else {
                //println!("skipping {}", line);
            }
            cur_pos += line.len() as u64;
            if cur_pos >= self.len {
                break;
            }
        }
        line_timestamp - self.b_time.timestamp()
    }

    fn call_inner(mut self, args: (u64,)) -> i64 {
        self.call_inner_mut(args)
    }
}

impl<'a, R: 'a + Read + Seek> FnOnce<(u64,)> for FilePredicate<'a, R> {
    type Output = i64;
    extern "rust-call" fn call_once(self, pos: (u64,)) -> i64 {
        self.call_inner(pos)
    }
}

impl<'a, R: 'a + Read + Seek> FnMut<(u64,)> for FilePredicate<'a, R> {
    extern "rust-call" fn call_mut(&mut self, pos: (u64,)) -> i64 {
        self.call_inner_mut(pos)
    }
}

fn print_usage(opts: Options) {
    let program = "timecutrust";
    let brief = format!("Usage: {} [options] 'beg-time' file", program);
    print!("{}", opts.usage(&brief));
    println!("examples:");
    println!("timecutrust '[2015-12-28 20:37:25]' ./sample_g2.log");
}


fn main() {
    let args: Vec<String> = env::args().collect();

    let mut opts = Options::new();
    opts.optopt("r", "regexp", "regular expresrion", "REGULAR_EXPRESSION");
    opts.optflag("h", "help", "print this help menu");
    opts.optflag("n", "nginx", "run on nginx access.log with pattern like '[28/Dec/2015:06:26:08'");

    let matches = match opts.parse(&args[1..]) {
        Ok(m) => { m }
        Err(f) => { panic!(f.to_string()) }
    };

    if matches.opt_present("h") {
        print_usage(opts);
        return;
    }

    if matches.free.len() < 1 {
        print_usage(opts);
    } else {
        let mut re = matches.opt_str("r")
            .unwrap_or(r"^\[(?P<year>\d{4})-(?P<month>\d{2})-(?P<day>\d{2})\s(?P<hour>\d{2}):(?P<minute>\d{2}):(?P<second>\d{2})\]"
                       .to_string());
        if matches.opt_present("n") {
            re = r"^\[(?P<day>\d{2})/(?P<monthname>\p{L}*)/(?P<year>\d{4}):(?P<hour>\d{2}):(?P<minute>\d{2}):(?P<second>\d{2})".to_string();
        }
        match work_on_files(&matches.free[0], &matches.free[1], &re) {
            Ok(_) => return,
            Err(e) => {
                if e.kind() == ErrorKind::BrokenPipe {
                    return;
                } else {
                    println!("error parsing header: {}", e);
                }
            }
        };
    }
}

fn work_on_files<'a>(b: &'a str, f_name: &'a str, re_str: &'a str) -> Result<(), io::Error> {
    let f = try!(File::open(f_name));
    let meta = try!(fs::metadata(f_name));
    let file = BufReader::new(&f);
    println!("date regexp {}", re_str);
    work_pred(b, f_name, re_str, file, meta.len())
}

fn work_pred<'a, R: 'a + Read + Seek>(b: &'a str, f_name: &'a str, re_s: &'a str, file: BufReader<R>, len: u64) -> Result<(), io::Error> {
    let re = datetimes::init(re_s);
    let b_time =
        datetimes::parse(&re, b).expect(&format!("could not parse begging time {}", b))
            - Duration::milliseconds(1);
    let pred: FilePredicate<R> = FilePredicate::new(file, len, re, b_time);
    let start_pos = try!(get_start_pos(pred, len));
    work_end(f_name, start_pos)
}

fn get_start_pos<'a, R: 'a + Read + Seek>(mut pred: FilePredicate<'a, R>, len: u64) -> Result<(u64), io::Error> {
    let start_pos = binary_search::binary_search(0, len, & mut pred);
    Ok(start_pos)
}

fn work_end(f_name: &str, start_pos: u64) -> Result<(), io::Error> {
    let f = try!(File::open(f_name));
    let mut file = BufReader::new(&f);
    let stdout = io::stdout();
    let mut out = stdout.lock();
    file.seek(SeekFrom::Start(start_pos)).unwrap();
    for r_line in file.lines() {
        let line = r_line.unwrap();
        try!(out.write_fmt(format_args!("{}\n", line)));
    }
    Ok(())
}


